processor/transformprocessor/internal/metrics/metrics.go

```diff
@@ -0,0 +1,941 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/metrics"
+import (
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/common"
+)
+
+type metricTransformContext struct {
+	dataPoint interface{}
+	metric    pmetric.Metric
+	il        pcommon.InstrumentationScope
+	resource  pcommon.Resource
+}
+
+func (ctx metricTransformContext) GetItem() interface{} {
+	return ctx.dataPoint
+}
+
+func (ctx metricTransformContext) GetInstrumentationScope() pcommon.InstrumentationScope {
+	return ctx.il
+}
+
+func (ctx metricTransformContext) GetResource() pcommon.Resource {
+	return ctx.resource
+}
+
+func (ctx metricTransformContext) GetDescriptor() interface{} {
+	return ctx.metric
+}
+
+// pathGetSetter is a getSetter which has been resolved using a path expression provided by a user.
+type pathGetSetter struct {
+	getter common.ExprFunc
+	setter func(ctx common.TransformContext, val interface{})
+}
+
+func (path pathGetSetter) Get(ctx common.TransformContext) interface{} {
+	return path.getter(ctx)
+}
+
+func (path pathGetSetter) Set(ctx common.TransformContext, val interface{}) {
+	path.setter(ctx, val)
+}
+
+func ParsePath(val *common.Path) (common.GetSetter, error) {
+	return newPathGetSetter(val.Fields)
+}
+
+func newPathGetSetter(path []common.Field) (common.GetSetter, error) {
+	switch path[0].Name {
+	case "resource":
+		if len(path) == 1 {
+			return accessResource(), nil
+		}
+		switch path[1].Name {
+		case "attributes":
+			mapKey := path[1].MapKey
+			if mapKey == nil {
+				return accessResourceAttributes(), nil
+			}
+			return accessResourceAttributesKey(mapKey), nil
+		}
+	case "instrumentation_scope":
+		if len(path) == 1 {
+			return accessInstrumentationScope(), nil
+		}
+		switch path[1].Name {
+		case "name":
+			return accessInstrumentationScopeName(), nil
+		case "version":
+			return accessInstrumentationScopeVersion(), nil
+		}
+	case "descriptor":
+		if len(path) == 1 {
+			return accessDescriptor(), nil
+		}
+		switch path[1].Name {
+		case "metric_name":
+			return accessDescriptorMetricName(), nil
+		case "metric_description":
+			return accessDescriptorMetricDescription(), nil
+		case "metric_unit":
+			return accessDescriptorMetricUnit(), nil
+		case "metric_type":
+			return accessDescriptorMetricType(), nil
+		case "metric_aggregation_temporality":
```

Comment:
We use the descriptor both for the upper level fields and the ones shared among different types of instruments. Seems reasonable to me.